translations_parser/parser.py

```diff
@@ -0,0 +1,251 @@
+from dataclasses import dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import List
+
+import yaml
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="[%(levelname)s] %(message)s",
+)
+logger = logging.getLogger(__name__)
+
+HEADER_RE = re.compile(r"(?<=\[)(?P<value>.+?)\] ")
+VALIDATION_RE = re.compile(r"Ep\.[ :]+(?P<ep>\d+)[ :]+Up\.[ :]+(?P<up>\d+)[ :]+(?P<key>[\w-]+)[ :]+(?P<value>[\d\.]+)")
+TRAINING_RE = re.compile(
+    r"Ep\.[ :]+(?P<epoch>\d+)[ :]+"
+    r"Up\.[ :]+(?P<up>\d+)[ :]+"
+    r"Sen\.[ :]+(?P<sen>[\d,]+)[ :]+"
+    r"Cost[ :]+(?P<cost>[\d.]+)[ :]+"
+    r"Time[ :]+(?P<time>[\d\.]+s)[ :]+"
+    r"(?P<rate>[\d\.]+ words\/s)[ :]+"
+    r"gNorm[ :]+(?P<gnorm>[\d\.]+)"
+)
+# Expected version of Marian for a clean parsing
+MARIAN_MAJOR, MARIAN_MINOR = 1, 10
+
+
+@dataclass
+class TrainingEpoch:
+    epoch: str
+    up: str
+    sen: str
+    cost: str
+    time: int
+    rate: str
+    gnorm: str
+
+
+@dataclass
+class ValidationEpoch:
+    epoch: str
+    up: str
+    chrf: str
+    ce_mean_words: str
+    bleu_detok: str
+
+
+@dataclass
+class TrainingLog:
+    """Results from the parsing of a training log file"""
+
+    # Marian information
+    info: dict
+    # Runtime configuration
+    configuration: dict
+    training: List[TrainingEpoch]
+    validation: List[ValidationEpoch]
+    # Dict of log lines indexed by their header (e.g. marian, data, memory)
+    logs: dict
+
+
+class TrainingParser:
+    def __init__(self, logs_iter):
+        # Iterable reading logs lines
+        self.logs_iter = logs_iter
+        self.parsed = False
+        self.config = {}
+        self.indexed_logs = defaultdict(list)
+        self.training = []
+        # Dict mapping (epoch, up) to values parsed on multiple lines
+        self.validation_entries = defaultdict(dict)
+        # Marian exection data
+        self.version = None
+        self.version_hash = None
+        self.release_date = None
+
+    def get_headers(self, line):
+        """
+        Returns a list of tuples representing the headers of a log line
+        and the position of the last index
+        """
+        matches = list(HEADER_RE.finditer(line))
+        if not matches:
+            return ((), None)
+        return ([tuple(m.group("value").split()) for m in matches], matches[-1].span()[-1])
+
+    def check_task_timestamp_header(self, values):
+        """
+        Check a header value matching ('task', <timestamp>)
+        and return the deduced timestamp
+        """
+        if not values or len(values) != 2:
+            return
+        base, timestamp = values
+        if base != "task":
+            return
+        return datetime.fromisoformat(timestamp.rstrip("Z"))
+
+    def parse_training_log(self, headers, text):
+        match = TRAINING_RE.match(text)
+        if not match:
+            return
+        self.training.append(TrainingEpoch(**match.groupdict()))
+
+    def parse_validation_log(self, headers, text):
```

Comment:
Do not store the units `s` or `word/s`: they should be in the regex, and not present in the CSV (it's harder to use otherwise)